### PR TITLE
Forking support For Span Processors and Exporters

### PR DESCRIPTION
**Pull request by snehilchopra**
This PR adds forking support for span processors, as per #85.

The implementation is quite similar to the [Fork Aware Interface](https://github.com/lightstep/lightstep-tracer-cpp/blob/master/src/recorder/fork_aware_recorder.h) implementation in the lightstep-tracer-cpp repo.

The idea is to expose an interface to allow all span processors the ability to register fork handlers using the [OT C++ portable wrapper](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/sdk/src/common/platform/fork.h) for [pthread_atfork](https://www.ibm.com/support/knowledgecenter/en/SSLTBW_2.2.0/com.ibm.zos.v2r2.bpxbd00/pthreadatfork.htm).

### CODE REVIEW

File: sdk/test/trace/batch_span_processor_test.cc [hunk 1]
```diff
@@ -291,4 +304,71 @@ TEST_F(BatchSpanProcessorTestPeer, TestScheduleDelayMillis)
...
+    EXPECT_TRUE(is_export_completed->load());
+    EXPECT_EQ(3, spans_received->size());
+
+    batch_processor->Shutdown();
+    // End child process
+    std::exit(0);
+  }
+
+  // We only created one span in the parent process
+  EXPECT_EQ(1, spans_received->size());
```
Comment (snehilchopra) [inline sdk/test/trace/batch_span_processor_test.cc:364]: All the `EXPECT` statements pass (when I run it locally on my machine), but I believe the CircleCI checks are not  checking the `EXPECT` statements in the child process.

Adding a shared memory segment (using `mmap`) to make sure that CircleCI checks whether or not the tests in the child process pass.

File: sdk/include/opentelemetry/sdk/trace/fork_aware_span_processor.h [hunk 2]
```diff
@@ -0,0 +1,88 @@
+// The MIT License (MIT)
+
+// Copyright (c) 2015-2016 LightStep
```
Comment (maxgolov) [inline sdk/include/opentelemetry/sdk/trace/fork_aware_span_processor.h:3]: @pyohannes @reyang 
I have a licensing question here. The rest of OpenTelemetry code is Apache.. I believe we had a concern once adding dependency on other MIT-licensed libraries, such as MS Guidelines Support Library (which is also MIT btw).... Are we OK now to change the policy and allow adding other company -licensed contributions, with their respected copyrights that do not match the license here: https://github.com/open-telemetry/opentelemetry-cpp/blob/master/LICENSE

File: sdk/include/opentelemetry/sdk/trace/batch_span_processor.h [hunk 3]
```diff
@@ -19,7 +19,7 @@ namespace trace
  * This is an implementation of the SpanProcessor which creates batches of finished spans and passes
  * the export-friendly span data representations to the configured SpanExporter.
  */
-class BatchSpanProcessor : public SpanProcessor
+class BatchSpanProcessor : public ForkAwareSpanProcessor
```
Comment (maxgolov) [inline sdk/include/opentelemetry/sdk/trace/batch_span_processor.h:22]: Shouldn't `BatchSpanProcessor` remain a platform-agnostic class? Forking is a concept very foreign to Windows..

File: sdk/include/opentelemetry/sdk/trace/simple_processor.h [hunk 4]
```diff
@@ -14,7 +14,7 @@ namespace trace
  *
  * OnEnd and ForceFlush are no-ops.
  */
-class SimpleSpanProcessor : public SpanProcessor
+class SimpleSpanProcessor : public ForkAwareSpanProcessor
```
Comment (maxgolov) [inline sdk/include/opentelemetry/sdk/trace/simple_processor.h:17]: This doesn't seem right to be added to the `SimpleSpanProcessor`. Can this be implemented in its own separate class? For example, we are not explicitly covering support for Windows 10 Universal Platform `App::OnSuspending` (platform -specific feature) in a **Simple** processor. I think *Simple* span processor has to remain simple.

File: sdk/include/opentelemetry/sdk/trace/batch_span_processor.h (hunk 3 above)
Comment (snehilchopra) [reply sdk/include/opentelemetry/sdk/trace/batch_span_processor.h:22]: Per our conversation on gitter, should I instead make two separate classes? Something like `ForkAwareBatchSpanProcessor` and `ForkAwareSimpleSpanProcessor`?

This way we can leave it to the user to decide what to use

File: sdk/include/opentelemetry/sdk/trace/simple_processor.h (hunk 4 above)
Comment (snehilchopra) [reply sdk/include/opentelemetry/sdk/trace/simple_processor.h:17]: Please see the [comment](https://github.com/open-telemetry/opentelemetry-cpp/pull/274#discussion_r469422028) above

File: sdk/include/opentelemetry/sdk/trace/batch_span_processor.h (hunk 3 above)
Comment (snehilchopra) [reply sdk/include/opentelemetry/sdk/trace/batch_span_processor.h:22]: Also, do you prefer any specific directory structure wherein we can store all ForkAware processors in one folder, and all the ForkAware exporters in another folder?

Comment (snehilchopra) [reply sdk/include/opentelemetry/sdk/trace/batch_span_processor.h:22]: Also, I had one general comment regarding inheriting from the `ForkAwareSpanProcessor` class. The class primarily just exposes three fork handlers that any custom processor can implement as required. 

Even if a processor (inheriting from the `ForkAwareSpanProcessor`) is used on a non-linux OS, there would not be essentially any breaks in the code, meaning that they would still function properly.

As far as registering fork handlers is concerned, that's already handled [here](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/sdk/src/common/CMakeLists.txt#L3) on a platform basis. You can also see [here](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/sdk/src/common/platform/fork_windows.cc) that no fork handlers are registered on the Windows platform.

**Comment by snehilchopra** [comment]
@reyang @maxgolov @pyohannes 
Added forking support for exporters as well in this PR because it wasn't too much of a clutter.

In addition, both the `ForkAware` processors AND exporters get tested in the same files, namely `fork_aware_batch_span_processor_test.cc` and `fork_aware_simple_processor_test.cc`. Both files have mock processors and mock exporters, which in turn are fork aware. So it becomes easier to test by registering fork handlers for both the processor and exporter in the same test suite.

**Comment by alolita** [comment]
Triaged Dec 7 2020: @maxgolov will sync w @pyohannes to figure out next steps.

**Pull request closed without merging**